### PR TITLE
[Dialog][Popover][DropdownMenu] Adjust non-modal auto focus behaviour

### PR DESCRIPTION
**Pull request by andy-hook**
Previously non-modal auto focus was not covering use of the close button. This change applies auto focus for all interactions except when pointer down outside.

### CODE REVIEW

File: packages/react/dropdown-menu/src/DropdownMenu.tsx
Comment (andy-hook) [inline packages/react/dropdown-menu/src/DropdownMenu.tsx:233]: We need to reset the ref here between closes, this change previously assumed it would only ever be closed via `escape` key. Now it will reset after every close regardless.

File: packages/react/dialog/src/Dialog.tsx
Comment (jjenzz) [inline packages/react/dialog/src/Dialog.tsx:269]: if you fancy being generous haha, nit 🙈 :

```suggestion
        onPointerDownOutside={composeEventHandlers(
          props.onPointerDownOutside,
          () => (isPointerDownOutsideRef.current = true),
          { checkForDefaultPrevented: false }
        )}
```

File: packages/react/popover/src/Popover.tsx
Comment (jjenzz) [inline packages/react/popover/src/Popover.tsx:280]: same nit 🙈 :

```suggestion
        onPointerDownOutside={composeEventHandlers(
          props.onPointerDownOutside,
          () => (isPointerDownOutsideRef.current = true),
          { checkForDefaultPrevented: false }
        )}
```

**Review by jjenzz — DISMISSED** [review]

**Comment by jjenzz** [comment]
❤️

File: packages/react/dropdown-menu/src/DropdownMenu.tsx
Comment (benoitgrelard) [reply packages/react/dropdown-menu/src/DropdownMenu.tsx:233]: This is one of those cases where we need to make sure they can't cancel that bit of the behaviour then right? Because right now, if they prevent default to run their own focus, the ref won't be reset.

File: packages/react/dialog/src/Dialog.tsx
Comment (benoitgrelard) [inline packages/react/dialog/src/Dialog.tsx:250]: This one is never getting reset in `DialogContentNonModal` or `PopoverContentNonModal`?

Comment (andy-hook) [reply packages/react/dialog/src/Dialog.tsx:250]: They are reset during teardown / remount. `DropdownMenuRootContent` is unusual because this is handled one level lower by `Menu`, which means the top level state persists unless we reset manually.

File: packages/react/dropdown-menu/src/DropdownMenu.tsx
Comment (andy-hook) [reply packages/react/dropdown-menu/src/DropdownMenu.tsx:233]: I think that depends, the current assumption makes this moot because if we `preventDefault` on `onCloseAutoFocus` then we're opting out of the built in focus handling all together right? or is the expectation that the user is cancelling the default behaviour of **the handler**, and that the conditions with which it is run should persist? e.g. the `onCloseAutoFocus` handler itself is subject to pointer conditions, not just the focus behaviour within it.

File: packages/react/dialog/src/Dialog.tsx
Comment (benoitgrelard) [reply packages/react/dialog/src/Dialog.tsx:250]: Teardown/remount isn't enough though, because you can have a pointer down but no close (so no teardown) for example when preventing default in `onPointerDownOutside` or in `onInteractOutside`.

File: packages/react/dropdown-menu/src/DropdownMenu.tsx
Comment (benoitgrelard) [reply packages/react/dropdown-menu/src/DropdownMenu.tsx:233]: Well I think it's more about ensuring the "integrity" of the system. That's kinda when we need to think about which bit they can or cannot prevent. ie. here it's important that it gets reset properly when it's actually closed.

File: packages/react/dialog/src/Dialog.tsx
Comment (andy-hook) [reply packages/react/dialog/src/Dialog.tsx:250]: Good point and made me question another assumption, given the the case you describe we should probably be resetting when interacting _inside_ as well.

Comment (benoitgrelard) [reply packages/react/dialog/src/Dialog.tsx:250]: It's really tricky, because basically what you want is reset it if there was a pointer down that didn't result in a close. But you don't really know this yet.

Comment (andy-hook) [reply packages/react/dialog/src/Dialog.tsx:250]: Yeah.. the simplest way I could see was to explicitly handle ref clearing https://github.com/radix-ui/primitives/pull/799/commits/b74b6cfed252e009e5e9f952bd50cd2bb8519027 it's a bit verbose though.

Comment (benoitgrelard) [reply packages/react/dialog/src/Dialog.tsx:250]: Yeah I don't love it haha, gonna have to think some more about this.
Any ideas @jjenzz?

Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx:250]: I'm not sure I'm fully understanding the issue. How do I replicate from within the `modality` branch?

- Open dialog
- Press close button
- Trigger doesn't auto focus?

I've just tried these steps in the `modality` branch and the trigger focuses fine so I'm missing something here for sure.

Comment (andy-hook) [reply packages/react/dialog/src/Dialog.tsx:250]: It should focus if you close via the button. Try opening non-modal with keyboard and closing with return / space, it should focus the trigger but it doesn't

Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx:250]: I'm probably missing something but would this suffice?

image


- Preventing default in `onPointerDownOutside` or `onFocusOutside` would prevent dismissing and `onInteractOutside` so would never set the ref to `true`.
- Dismissing via outside interaction only would set it to `true` which would close it and bring it back into it's default `false` state upon remount.

Comment (benoitgrelard) [reply packages/react/dialog/src/Dialog.tsx:250]: > … and `onInteractOutside` so would never set the ref to true.

That's not the case right? `onInteractOutside` will run even if you call `preventDefault` in `onPointerDownOutside` or `onFocusOutside`.

Comment (benoitgrelard) [reply packages/react/dialog/src/Dialog.tsx:250]: But now it's making me think, perhaps it's as simple as resetting it in `onInteractOutside` when `!event.defaultPrevented`?

Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx:250]: > `onInteractOutside` will run even if you call `preventDefault` in `onPointerDownOutside` or `onFocusOutside`.

Nope, which is really baffling to me but it doesn't. Because yah, I originally did what you suggested and console.logged `event.defaultPrevented` because I wanted to check that, but it never logged ...

Comment (benoitgrelard) [reply packages/react/dialog/src/Dialog.tsx:250]: hmmm

image

The only thing behind the `defaultPrevented` check is `onDismiss`…

Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx:250]: If we want to be sure though, seems this should be all we need:

```diff
- else  isInteractOutsideRef.current = true;
+ else if (!event.defaultPrevented) isInteractOutsideRef.current = true;
```

Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx:250]: Yep, I have no idea what's going on there. It's made me worry that we might have a load of logic elsewhere that isn't running 😅

Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx:250]: If I change it to this then it calls it fine, but with the `event` param it doesn't

image

Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx:250]: Looks like there is a bug somewhere which we can fix separately, and we should definitely do the `!event.defaultPrevented` check in the meantime because `onInteractOutside` should be called.

Comment (andy-hook) [reply packages/react/dialog/src/Dialog.tsx]: Did I open another can of worms here? haha. I've updated from the pro tips, is this what you had in mind? https://github.com/radix-ui/primitives/pull/799/commits/dfe047989d4ac612b3d1836fb9853d60415620b8

File: packages/react/dropdown-menu/src/DropdownMenu.tsx
Comment (andy-hook) [inline packages/react/dropdown-menu/src/DropdownMenu.tsx:253]: Do we foresee any issue with this? was easier to move it from `onPointerDownOutside` as we have to account for both modes in this comp.

File: packages/react/dialog/src/Dialog.tsx
Comment (benoitgrelard) [reply packages/react/dialog/src/Dialog.tsx]: haha I "think" that's correct, but man this is a real slippery area 😀 
I'm gonna have to take a deeper look at it and actually play with it myself a bit to get my head around it properly.

File: packages/react/dialog/src/Dialog.tsx
Comment (benoitgrelard) [inline packages/react/dialog/src/Dialog.tsx:265]: Man, this is all getting so confusing because of our `checkForDefaultPrevented` thing haha. Next step will really be to do something about this because it's melting my brain 😀

Here for example, it feels weird to see a check for `!event.defaultPrevented` and also have `checkForDefaultPrevented: false` 😀

Also, the placement of our own `event.preventDefault()` bothers me…
- I understand we need to do it in **ALL** cases because either:
  - we don't want to run any focus at all (ie. you've clicked somewhere outside, so natural focus should occur)
  - we want explicit focus to trigger so we prevent the default close auto focus
- So, instinctively I want to put it as the first line at the beginning of the handler, but we can't because we immediately check for `!event.defaultPrevented` (because we need to know if it's been prevented by the user, not us)

As a test I tried to see how it would look like if we didn't use `composeEventHandlers` and ended with this which definitely makes me think we should move away from it soon because I find it way clearer:

```tsx
onCloseAutoFocus={(event) => {
  props.onCloseAutoFocus?.(event);
  const userPrevented = event.defaultPrevented;

  if (!userPrevented) {
    event.preventDefault();
    if (!isInteractOutsideRef.current) context.triggerRef.current?.focus();
  }

  isInteractOutsideRef.current = false;
}}
```

With the current way we end up with a double negative sort of situation, I find it super hard to reason about, especially because for example our own `event.preventDefault()` is run even if the user has already prevented. I know it doesn't make any difference, but it makes it quite confusing when I read it.

What do we all think about this? Do we wait? or do we bite the bullet and start here with this change before updating everywhere later?

Comment (andy-hook) [reply packages/react/dialog/src/Dialog.tsx:265]: I'm glad you bring this up because this whole thing often makes my head spin and I find myself wondering whether I'm just dense or if this approach needs adjustment.

In your example, will the consumer setting

```jsx
onCloseAutoFocus={(event) => event.preventDefault()}
```

Revert the auto focus behaviour to the implicit, last focused element thing?

Comment (benoitgrelard) [reply packages/react/dialog/src/Dialog.tsx:265]: No if the user prevent defaults, no focus behaviour will happen because the check is done in `FocusScope` and we pass it down: `onUnmountAutoFocus={onCloseAutoFocus}`.

Comment (andy-hook) [reply packages/react/dialog/src/Dialog.tsx:265]: My 2c is that your suggestion makes A LOT more sense and I would prefer to go with it here. I also support incrementally moving away from the utils if we agree that's the right thing to do.

Comment (benoitgrelard) [reply packages/react/dialog/src/Dialog.tsx:265]: Yeah at this point I think I'd also say we should do it incrementally as it seems pointless to have those things use it even though we all feel it's making things worse and harder to understand.

Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx:265]: Aw man, I totally agree with all of this and **MUCH** prefer your suggestion Benoit. I don't think we need to make huge sweeping codebase changes if we decide to change direction, I think it's fine to start here and avoid `composeEventHandlers` going forward and have a separate task to update legacy stuff.

Comment (benoitgrelard) [reply packages/react/dialog/src/Dialog.tsx:265]: Yay for all agreeing 🎉 
Let's goooo!

Comment (andy-hook) [reply packages/react/dialog/src/Dialog.tsx]: I've tweaked this for non-modal for now https://github.com/radix-ui/primitives/pull/799/commits/cdc24481073ed22d40c98bc816316ec58a0f719d will update the others in those files separately if we're good.

Question on moving away from the utils, in the common case does this mean we'll just be adopting an inline condition in place?

e.g.

```jsx
onClick={composeEventHandlers(onClick, () => {
  ...
})}
      
// Will become

onClick={(event) => {
  props.onClick?.(event)
  const userPrevented = event.defaultPrevented

  if (!userPrevented) {
   ...
  }
}}
```

Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx]: Yeah although I'm find with just doing:

```jsx
onClick={(event) => {
  props.onClick?.(event)
  if (!event.defaultPrevented) {
   ...
  }
}}
```

It's not necessarily "userPrevented". We could prevent things internally via a capture event or something for example.

File: packages/react/dialog/src/Dialog.tsx
Comment (jjenzz) [inline packages/react/dialog/src/Dialog.tsx:269]: Was there a reason we didn't go with this?

```suggestion
        onCloseAutoFocus={(event) => {
          props.onCloseAutoFocus?.(event);
          if (!event.defaultPrevented && !isInteractOutsideRef.current) {
			context.triggerRef.current?.focus();
          }

          isInteractOutsideRef.current = false;
          event.preventDefault();
        }}
```

The current one confused me because `preventDefault` is always happening either way. Either they are or we are so we may as well `event.preventDefault()` all the time. It didn't make sense why that had to be part of the conditional.

File: packages/react/dialog/src/Dialog.tsx
Comment (jjenzz) [inline packages/react/dialog/src/Dialog.tsx:283]: again, I'm fine with just `else if (!event.defaultPrevented)` here because it's not necessarily the user preventing it.

**Review by jjenzz — APPROVED** [review]
some nits, but not gonna hold up the PR for them. Rest looks good :+1:

File: packages/react/dialog/src/Dialog.tsx
Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx:269]: In fact, shouldn't we only be preventing default if we're focusing ourselves? so it would be in **both** conditionals:

```jsx
        onCloseAutoFocus={(event) => {
          props.onCloseAutoFocus?.(event);
          if (!event.defaultPrevented && !isInteractOutsideRef.current) {
            event.preventDefault();
            context.triggerRef.current?.focus();
          }
          isInteractOutsideRef.current = false;
        }}
```

?

Comment (andy-hook) [reply packages/react/dialog/src/Dialog.tsx:269]: Mainly for the ordering reason in Benoit's comment https://github.com/radix-ui/primitives/pull/799#discussion_r678099342 if we take it out of the condition then we need to ensure it comes after.

File: packages/react/dialog/src/Dialog.tsx
Comment (andy-hook) [reply packages/react/dialog/src/Dialog.tsx:283]: Yeah I'm on board with this 👍

File: packages/react/dialog/src/Dialog.tsx
Comment (andy-hook) [reply packages/react/dialog/src/Dialog.tsx:269]: > In fact, shouldn't we only be preventing default if we're focusing ourselves?

Ah I didn't see this comment before I added mine, yeah I think you're right!

**Review by jjenzz — APPROVED** [review]
Code-wise looks good to me but I'll leave it to Benoit to make sure it's all good behaviour-wise since he has a much deeper understanding than me here 😅

File: packages/react/dialog/src/Dialog.tsx
Comment (benoitgrelard) [reply packages/react/dialog/src/Dialog.tsx]: No, that's not right, we need to prevent **all the time** not just when we're focusing something manually. Like I explained here: https://github.com/radix-ui/primitives/pull/799#discussion_r678099342

> I understand we need to do it in ALL cases because either:
> - **we don't want to run any focus at all (ie. you've clicked somewhere outside, so natural focus should occur)**
> - we want explicit focus to trigger so we prevent the default close auto focus

File: packages/react/dropdown-menu/src/DropdownMenu.tsx
Comment (benoitgrelard) [inline packages/react/dropdown-menu/src/DropdownMenu.tsx:233]: `DropdownMenu.Root` can be modal or non-modal right?
So in this case this is correct here t only `preventDefault` if we focus ourselves.

File: packages/react/dropdown-menu/src/DropdownMenu.tsx
Comment (benoitgrelard) [inline packages/react/dropdown-menu/src/DropdownMenu.tsx:234]: This check is already in the above conditional.

File: packages/react/popover/src/Popover.tsx
Comment (benoitgrelard) [inline packages/react/popover/src/Popover.tsx:274]: That one is wrong too I think, same as non modal dialog.

File: packages/react/dropdown-menu/src/DropdownMenu.tsx
Comment (benoitgrelard) [inline packages/react/dropdown-menu/src/DropdownMenu.tsx:254]: I'm still struggling to follow the double negatives a bit, like we set a `isInteractOutside` which is about what "happened" and set it to the negative of something.

Then later we do `if (!isInteractOutside…)`

I feel like it would be easier for the ref we track to be about the "what" rather than the "how", ie. something like `shouldReturnFocus`.

Can we go on a call and pair quickly to see if it would help?

**Review by benoitgrelard — APPROVED** [review]

File: packages/react/dialog/src/Dialog.tsx
Comment (andy-hook) [reply packages/react/dialog/src/Dialog.tsx]: @benoitgrelard is on the money here, it always needs to happen in non-modal to avoid focus fighting on outside clicks etc. https://github.com/radix-ui/primitives/pull/799/commits/fb6e32cca44c9a271a9ada0db4d75713e230e308 this whole thing got a little confusing after a while 😅

**Comment by andy-hook** [comment]
@jjenzz – Me and @benoitgrelard had a call on this and made some readability changes plus reverted some wrong assumptions I had with the last round of changes. Going to merge now and we can discuss more in the main modality branch if needed.

File: packages/react/dialog/src/Dialog.tsx
Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx]: Ah of course... wow, yeah, my head is spinning 😅

Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx]: On the whole "coming after" thing. What's wrong with this? In the olden days this was actually the recommended place to put it to ensure progressive enhancement didn't break if there was a JS error in the handling logic.

Comment (jjenzz) [reply packages/react/dialog/src/Dialog.tsx]: See this example where when JS fails, the app still works. **if the event is prevented after** https://codepen.io/pen/?editors=1010

File: packages/react/context-menu/src/ContextMenu.tsx
Comment (jjenzz) [inline packages/react/context-menu/src/ContextMenu.tsx:217]: @andy-hook I saw your comment about further discussing in the `modality` branch but i find it's easier to follow if we keep all discussion related to this change in one place.

So, on this one, curious why we no longer focus the trigger here when we prevent default? We had to do it explicitly in `main` to fix `autoFocus` issues.

Comment (andy-hook) [reply packages/react/context-menu/src/ContextMenu.tsx:217]: This change was to match the non-modal behaviour we have in the other components. That is that when clicking outside in non-modal, focus shouldn't be returned.

Not familiar with the issue you mentioned, do you have a link to it? if this will cause issues would be good to consider it.

@benoitgrelard probably has a better explanation as he originally brought this to my attention.

Comment (jjenzz) [reply packages/react/context-menu/src/ContextMenu.tsx:217]: Oh wait, this is `ContextMenu` and they can't `autoFocus` a menu item so it's okay to fallback onto `FocusScope`s auto focus behaviour in this case :+1:

This was the original `autoFocus` issue if you're curious https://github.com/radix-ui/primitives/issues/518. I went through all our components and explicitly focused the trigger on unmount but looks like I also decided against applying it to `ContextMenu` at the time.

**Pull request merged**
